### DISCUSSION TITLE
The best way to show a timer while processing something asynchronously?

### DISCUSSION THREAD

**Discussion by amano-kenji** [jtdaugherty/brick#486] [Q&A] [answered]
Calculating statistics for many database table rows can take a few minutes.

I want to show a timer to users. The timer should be updated each second.

When the calculation is complete, the calcaulation task should end the timer and send an event to the application.

Have you done something like this before?

**Comment by jtdaugherty** [comment]
Yes, I've done this sort of thing many times. The [CustomEventDemo](https://github.com/jtdaugherty/brick/blob/master/programs/CustomEventDemo.hs) program provided with `brick` provides some scaffolding for doing this sort of thing, but the idea is:

* Create a custom event type with constructors for the events that should trigger an application state update / redraw. In your case that might be something like `data MyEvent = TimerUpdated Double | CalculationFinished`.
* Create a custom *request* type used to queue up work to do. In your case that might be something like `data Request = CalculateStats DatabaseQuery` or similar.
* Spawn a thread to do the work, taking requests from a `TChan` and sending responses back over a `BChan` back to the Brick application event loop. To do this you'll need to create a `BChan` and use `customMain` to run Brick.

**Comment by amano-kenji** [comment, marked as answer]
I found a solution.

```
now <- getCurrentTime
stopTimer <- newEmptyTMVarIO
timerStopped <- newEmptyTMVarIO
let calcText old new = T.unlines
      [ "Calculating statistics..."
      , "If you want to stop this, kill this program"
      , "Elapsed time: " <> T.pack (formatTime defaultTimeLocale "%2h:%2M:%2S" $ diffUTCTime new old) ]
    timer = do
      now' <- getCurrentTime
      writeBChan respC $ AsyncEvent $ modify $ set #mode $ Processing $ calcText now now'
      threadDelay 1000000
      stop <- atomically $ tryTakeTMVar stopTimer
      case stop of
        Nothing -> timer
        Just () -> atomically $ putTMVar timerStopped ()
void $ forkIO timer
do the work that takes a long time
putTMVar stopTimer ()
takeTMVar timerStopped
writeBChan respC $ AsyncEvent $ change UI....
```

I can put this in an asynchronous worker thread. This can be executed in response to a specific asynchronous worker request.

`writeBChan respC $ AsyncEvent` sends an asynchronous event back to brick UI.
